File: 프로그래머스/3/132266. 부대복귀/부대복귀.py
from collections import deque

def solution(n, roads, sources, destination):
    answer = []
    
    graph = [[] for _ in range(n + 1)]
    visited = [-1] * (n + 1)
    for a, b in roads:
        graph[a].append(b)
        graph[b].append(a)
        
    q = deque([destination])
    visited[destination] = 0
    while q:
        v = q.popleft()
        for i in graph[v]:
            if visited[i] == -1:
                visited[i] = visited[v] + 1
                q.append(i)
    
    for i in sources:
        answer.append(visited[i])
    
    return answer

File: 프로그래머스/3/132266. 부대복귀/부대복귀.js
function solution(n, roads, sources, destination) {
    var answer = [];
    
    const MAP = Array.from({ length: n + 1 }, () => []);
    
    for (let road of roads) {
        let [a, b] = road;
        MAP[a].push(b);
        MAP[b].push(a);
    }
    
    let q = [destination];
    const visited = Array(n + 1).fill(-1);
    visited[destination] = 0;
    
    while (q.length) {
        let cur = q.shift();
        
        for (const next of MAP[cur]) {
            if (visited[next] === -1) { 
                q.push(next);
                visited[next] = visited[cur] + 1;
            }
        }
    }
    
    sources.forEach((s) => {
        answer.push(visited[s]);
    });
    
    return answer;
}
